shite.1
.Dd 2023
.Dt SHITE 1
.Os
.Sh NAME
.Nm shite
.Nd posix shell, static webpages generator
.Sh SYNOPSIS
.Nm
.Ar SRCDIR
.Ar DESTDIR
.Sh DESCRIPTION
.Nm
is a script to generate a static web page, optionally with blog and/or multiple
pages, by reading a bunch of markdown files and a simple config file.  Made to
facilitate blogging, writing and publication of online content. Since markdown
format is almost pure text, it can prevent the author from contact with ugly html
pages.  It is a very static program, offering little easy customization of the
final product. It will point all your pages to a "style.css" and "favicon.ico" file, that should be placed on the
main directory. With style.css you can customize your website, there is a working 
example on shite sources, just copy the file to the folder of your website.
.Sh ENVIRONMENT
.Ev MARKDOWN
enviroment variable is used to generate html pages.
.Sh FILES
.Pa SRCDIR
- where the site source is, specified by command flag
.Pp
.Pa SRCDIR/foo.md
- these markdown files on the mair dir will be linked as
pages on the naviagtion bar. As an example, this can be used
to create a contact's page, by creating a contact.md file.
.Pp
.Pa SRCDIR/foo.link
- link files. Each '.link' file will be a simple link in the navigation bar, poiting
to the site or file written on the first line.
.Pp
.Pa SRCDIR/shiterc
- the basic site configuration, where you set the title, subtitle, and footer text,
write these things in markdown, as done in the example.
.Pp
.Pa SRCDIR/BLOGDIR
- the root directory of your blog posts, this is set in shiterc, you chan chose any name.
If this folder do not exists, no blog will
be generated neither linked to the navigation bar.
.Pp
.Pa SRCDIR/BLOGDIR/'Category 1'/
- a folder which will hold your posts in a specific category. The folder name will be a section on blog index.
.Pp
.Pa SRCDIR/BLOGDIR/'Section Name'/n-post-name.md
- your blogpost into a specific section of your blog, written in markdown.
The number n on the name will determine the order the post will appear. 
The first two lines of the file must be the name and the date,
in this format:
.Bd -literal -offset left
#Title
1234/12/12
.Ed
.Pp
.Pa SRCDIR/BLOGDIR/'Section Name'/n-linkname.link
- this file will generate a blog entry on the specific section with a link to a file or webpage.
This is useful as an example for posting pdf files. The first two lines of this file have to be the title and date
as the markdown files. In the third line you add the link.
.Bd -literal -offset left
#Link title 
1234/12/12
https://www.link.com
.Ed
.Pp

.Sh EXAMPLES
You can start your site on top of the example site. Just run:
.Bd -literal -offset left
.Nm mkdir dest
.Nm Pa exemplesite/ dest/
.Nm cp exemplesite/style.css exemplesite/favicon.ico dest/
.Pp
And your site will be ready on dest/index.html
.Pp
You will need to copy the "style.css" and "favicon.ico" file to the destination 
dir only once.
.Sh SHITERC
Your site source
.Em must
have a shiterc file on it, this file will contain the "title", "subtitle", "footer"
and "blogdir" info of your site. You can add a file with some script (e.g. disquss)
on SRCDIR and point that file on shiterc with the variable "comment_script", this 
script will be on the footer of all your BLOGDIR pages. 
You can ignore BLOGDIR if you don't have one. You can set BLOGDIR to anything
you want the page with your articles to be called.
.Ed
.Sh SEE ALSO
.Xr smu 1
.Sh AUTHORS
.An Henrique Lengler Aq Mt hleng@tutanota.com
.Sh SECURITY CONSIDERATIONS
.Nm
should not delete any files, only overwrite used files.


shite.sh
#!/bin/sh
currentdir="$(pwd)"
html_header() { #$1 = css location, $2 favicon location, $3 index location
cat <<!__EOF__
<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=0.7">
		<link rel="stylesheet" href="$1">
		<link rel="icon" href="$2" type="image/x-icon">
		<title>$title - $subtitle</title>
	</head>
	<body>
		<header>
			<h1 id="title"><a href=$3>$title</a></h1>
			<h2 id="subtitle">&mdash; $subtitle</h2>
		</header>
		<ul id="nav">
			<li><a href=$3>home</a></li>
!__EOF__
}

# $1, print script, yes = 1, no = 0
html_footer() {
echo '		<footer>'
if [ "$1" ]; then
	cat "$srcdir"/"$comment_script"
fi
printf '\n'
cat <<!__EOF__
			<p> $footer
		</footer>
	</body>
</html>
!__EOF__
}

barentry() { # $1 = file, $2 = name
cat <<!__EOF__
			<li><a href="$1">$2</a></li>
!__EOF__
}

die() {
	echo "$@" >&2
	exit 1
}

# PARSE FLAGS AND FOLDERS
if [ "$1" = "-h" ] || [ "$1" = "" ]; then
	echo usage: "$0" srcdir destdir
	exit
fi

if [ "$2" = "" ]; then
	die "you must specify a destdir"
else
	srcdir="$1"
	destdir="$2"
fi

# test dirs
# src - allways needed
if [ ! -d "$srcdir" ]; then
	die "$srcdir isn't a directory, please check it"
else
	srcdir="$(cd "$srcdir" && pwd && cd "$currentdir")"
	echo source path: "$srcdir"
fi

# check shiterc
if [ -f "$srcdir"/shiterc ]; then
	echo 'found shiterc, parsing it'
	. "$srcdir"/shiterc
	echo parsed from "$srcdir""shiterc":
	echo "\ttitle: $title"
	echo "\tsubtitle: $subtitle"
	echo "\tfooter: $footer"
	echo "\tblogdir: $blogdir"
	echo "\tcomment script: $comment_script"
else
	die "didn't found shiterc, can't proceed without it"
fi

# blog
if [ -d "$srcdir"/"$blogdir" ]; then
	blog=1
	blogentries=0
else
	echo 'no blog, not building it'
	blog=0
fi

# check markdown
MARKDOWN="$(printenv MARKDOWN)"
if [ "$MARKDOWN" = "" ]; then
	die "couldn't build site, markdown processor not found: set MARKDOWN"
else
	if type "$MARKDOWN" >/dev/null 2>&1; then
		echo "markdown processor: $MARKDOWN"
		title="$(echo "$title" | "$MARKDOWN" | sed 's/<[^a>]*>//g')"
		subtitle="$(echo "$subtitle" | "$MARKDOWN" | sed 's/<[^a>]*>//g')"
		footer="$(echo "$footer" | "$MARKDOWN" | sed 's/<[^a>]*>//g')"
	else
		die "MARKDOWN isn't a suitable binary"
	fi
fi

# main
mkdir -p "$destdir"
destdir="$(cd "$destdir" && pwd && cd "$currentdir")"
echo "destination path: $destdir"

touch "$destdir"/index.html

header="$(mktemp shite-header.XXXXXX)"

echo "$destdir"/"$blogdir"

if [ "$blog" = 1 ]; then
	blogheader="$(mktemp -t shite-blogheader.XXXXXX)"
	if [ "$(ls -A "$srcdir"/"$blogdir")" ]; then
		blogfiles=1
		# create header for files on level deeper (on folders)
		blogheader2="$(mktemp -t shite-blogheader2.XXXXXX)"
	else
		blogfiles=0
	fi
fi

ls -1 "$srcdir" | while read -r file; do
	if [ "$blog" = 1 ] && [ "$blogentries" = 0 ]; then
		echo adding blog to headers
		mkdir -p "$destdir"/"$blogdir"
		html_header ../style.css ../favicon.ico ../index.html > \
			"$destdir"/"$blogdir"/index.html
		barentry "$blogdir"/index.html "$blogdir" >> "$header"
		barentry index.html "$blogdir" >> "$blogheader"
		if [ "$blogfiles" = 1 ]; then
			html_header ../../style.css ../../favicon.ico ../../index.html > \
			"$blogheader2"
			barentry ../index.html "$blogdir" >> "$blogheader2"
		fi
		blogentries=1
	fi

	case "$file" in
		*.md)
			filename="$(echo "$file" | sed 's/\.md//')"
			html_header style.css favicon.ico index.html > "$destdir"/"$filename".html
			if [ "$filename" != "index" ]; then
				barentry "$filename".html "$filename" >> "$header"
				if [ "$blog" = 1 ]; then
					barentry ../"$filename".html "$filename" >> "$blogheader"
				fi
				if [ "$blogfiles" = 1 ]; then
					barentry ../../"$filename".html "$filename" >> "$blogheader2"
				fi
			fi
			;;

		*.link)
			echo "link file found: $file"
			filename="$(echo "$file" | sed 's/\.link//')"
			url="$(sed 1q "$srcdir"/"$file")"
			barentry "$url" "$filename" >> "$header"
			if [ "$blog" = 1 ]; then
				barentry "$url" "$filename" >> "$blogheader"
			fi
			if [ "$blogfiles" = 1 ]; then
				barentry "$url" "$filename" >> "$blogheader2"
			fi
			;;
	esac
done 

ls -1 "$srcdir" | grep md | sed 's/\.md//' | while read -r file; do
	{
		cat "$header"
		echo '		</ul>'
		printf '\n\n<!--Begin markdown generated content-->\n\n'
		"$MARKDOWN" < "$srcdir"/"$file".md
		printf '\n\n<!--End markdown generated content-->\n\n'
		html_footer
	} >> "$destdir"/"$file".html
done

if [ "$blog" = 1 ]; then
	cat "$blogheader" >> "$destdir"/"$blogdir"/index.html
	echo '		</ul>' >> "$destdir"/"$blogdir"/index.html

	if [ "$blogfiles" = 1 ]; then
		echo '		<ul id="blogpost">' >> "$destdir"/"$blogdir"/index.html

		# read and link section by section
		find "$srcdir"/"$blogdir" -type d ! -name "$blogdir" | sed 's#.*/##' | \
		while read section; do
			echo "<h2 id=\"section\">$section</h2>" >> "$destdir"/"$blogdir"/index.html
			# if section have files
			if [ "$(ls -A "$srcdir"/"$blogdir"/"$section")" ]; then
				section_dest="$(echo "$section" | sed s/\ /_/g)"
				mkdir -p "$destdir"/"$blogdir"/"$section_dest"
				ls -1r "$srcdir"/"$blogdir"/"$section" | while read -r file; do
					case "$file" in
						*.md)
							file="$(echo $file | sed 's/\.md//')"
							posttitle="$(sed 1q "$srcdir"/"$blogdir"/"$section"/"$file".md | sed s/#//)"
							postdate="$(sed -n 2p "$srcdir"/"$blogdir"/"$section"/"$file".md)"
							echo "          <li>$postdate - <a href=\""$section_dest"/"$file".html\">$posttitle</a></li>" >> \
								"$destdir"/"$blogdir"/index.html
							touch "$destdir"/"$blogdir"/"$section_dest"/"$file".html
						{
							cat "$blogheader2"
							echo '		</ul>'
							echo "<h1 id=\"posttitle\">$posttitle</h1>"
							echo "<h2 id=\"postdate\">Written in: $postdate</h2>"
							printf '\n\n<!--Begin markdown generated content-->\n\n'
							sed 1,2d "$srcdir"/"$blogdir"/"$section"/"$file".md | "$MARKDOWN"
							printf '\n\n<!--End markdown generated content-->\n\n'
							if [ "$comment_script" ]; then
								html_footer yes
							fi
						} > "$destdir"/"$blogdir"/"$section_dest"/"$file".html
						echo "\tMARKDOWN: $section/$file"
						;;
						*.link)
							file="$(echo $file | sed 's/\.link//')"
							echo link file found: "$file"
							posttitle="$(sed 1q "$srcdir"/"$blogdir"/"$section"/"$file".link | sed s/#//)"
							postdate="$(sed -n 2p "$srcdir"/"$blogdir"/"$section"/"$file".link)"
							postlink="$(sed -n 3p "$srcdir"/"$blogdir"/"$section"/"$file".link)"
							echo LINK POST  $postlink
							echo "          <li>$postdate - <a href=\""$postlink"\">$posttitle</a></li>" >> \
								"$destdir"/"$blogdir"/index.html
						;;
					esac
				done
			else
				echo "<p>Nothing on this category</p>" >> "$destdir"/"$blogdir"/index.html
			fi
	done
		echo '		</ul>' >> "$destdir"/"$blogdir"/index.html
	else
		echo "<p id=\"warn\">There's nothing here, yet</p>" >> "$destdir"/"$blogdir"/index.html
	fi
	html_footer >> "$destdir"/"$blogdir"/index.html
	rm "$blogheader"
fi
	rm "$header"
	printf '\n'
	echo "site built"
	exit 0
